remote-storage-managers/s3/src/main/java/org/apache/kafka/rsm/s3/S3RemoteStorageManager.java

```diff
@@ -0,0 +1,215 @@
+        Objects.requireNonNull(endpointConfiguration, "endpointConfiguration must not be null");
+        this.endpointConfiguration = endpointConfiguration;
+    }
+
+    @Override
+    public void configure(final Map<String, ?> configs) {
+        Objects.requireNonNull(configs, "configs must not be null");
+
+        final S3RemoteStorageManagerConfig config = new S3RemoteStorageManagerConfig(configs);
+        this.bucket = config.s3BucketName();
+
+        AmazonS3ClientBuilder s3ClientBuilder = AmazonS3ClientBuilder.standard();
+        if (this.endpointConfiguration == null) {
+            s3ClientBuilder = s3ClientBuilder.withRegion(config.s3Region());
+        } else {
+            s3ClientBuilder = s3ClientBuilder.withEndpointConfiguration(endpointConfiguration);
+        }
+
+        // It's fine to pass null in here.
+        s3ClientBuilder.setCredentials(config.awsCredentialsProvider());
+
+        s3Client = s3ClientBuilder.build();
+        transferManager = TransferManagerBuilder.standard().withS3Client(s3Client).build();
+    }
+
+    @Override
+    public RemoteLogSegmentContext copyLogSegment(final RemoteLogSegmentId remoteLogSegmentId,
+                                                  final LogSegmentData logSegmentData) throws RemoteStorageException {
+        Objects.requireNonNull(remoteLogSegmentId, "remoteLogSegmentId must not be null");
+        Objects.requireNonNull(logSegmentData, "logSegmentData must not be null");
+
+        try {
+            final String logFileKey = s3Key(remoteLogSegmentId, logSegmentData.logSegment().getName());
+            log.debug("Uploading log file: {}", logFileKey);
+            final Upload logFileUpload = transferManager.upload(this.bucket, logFileKey, logSegmentData.logSegment());
+
+            final String offsetIndexFileKey = s3Key(remoteLogSegmentId, logSegmentData.offsetIndex().getName());
+            log.debug("Uploading offset index file: {}", offsetIndexFileKey);
+            final Upload offsetIndexFileUpload = transferManager.upload(this.bucket, offsetIndexFileKey, logSegmentData.offsetIndex());
+
+            final String timeIndexFileKey = s3Key(remoteLogSegmentId, logSegmentData.timeIndex().getName());
+            log.debug("Uploading time index file: {}", timeIndexFileKey);
+            final Upload timeIndexFileUpload = transferManager.upload(this.bucket, timeIndexFileKey, logSegmentData.timeIndex());
+
+            logFileUpload.waitForUploadResult();
+            offsetIndexFileUpload.waitForUploadResult();
+            timeIndexFileUpload.waitForUploadResult();
+
+            // TODO clean up in case of interruption
+
+            return new S3RemoteLogSegmentContext(
+                logSegmentData.logSegment().getName(),
+                logSegmentData.offsetIndex().getName(),
+                logSegmentData.timeIndex().getName()
+            );
+        } catch (final SdkClientException e) {
```

Comment:
Also catch any exception so as to make sure all exceptions propagated by this method are of type `RemoteStorageException`, that helps caller dealing with failure modes as a single type of exception can be expected. Additionally, contextual information can be added to the `RemoteStorageException` in the future to better categorize the type of error (e.g. invalid configuration, unavailability of the remote storage, etc.), again helping with failure handling in the RSM. To that end, the categories of errors offered to implementations should be defined in the framework itself and not be storage-implementation specific. The example of HDFS and S3 storage allows to collect types failures from which the least common denominators can be extracted.